src/imageio/format/jxl.c

```diff
@@ -0,0 +1,749 @@
+      }
+
+      if(!write_icc)
+      {
+        color_encoding.primaries = JXL_PRIMARIES_CUSTOM;
+        color_encoding.primaries_red_xy[0] = primaries.Red.x;
+        color_encoding.primaries_red_xy[1] = primaries.Red.y;
+        color_encoding.primaries_green_xy[0] = primaries.Green.x;
+        color_encoding.primaries_green_xy[1] = primaries.Green.y;
+        color_encoding.primaries_blue_xy[0] = primaries.Blue.x;
+        color_encoding.primaries_blue_xy[1] = primaries.Blue.y;
+        // TODO: safe to simply put pipe->icc_intent here?
+        color_encoding.rendering_intent = JXL_RENDERING_INTENT_PERCEPTUAL;
+
+        JxlEncoderSetColorEncoding(encoder, &color_encoding);
+      }
+    }
+  }
+
+  if(write_icc)
+  {
+    fprintf(stderr, "JXL: Could not generate color encoding data, falling back to ICC binary\n");
+
+    cmsUInt32Number size = 0;
+    if(cmsSaveProfileToMem(out_profile, NULL, &size))
+    {
+      unsigned char *buf = malloc(size);
+      if(!buf || !cmsSaveProfileToMem(out_profile, buf, &size)) goto end;
+
+      JXL_ASSERT(JxlEncoderSetICCProfile(encoder, buf, size));
+
+      free(buf);
+    }
+    else
+    {
+      fprintf(stderr, "JXL: Error writing ICC data\n");
+      goto end;
+    }
+  }
+
+  JxlBasicInfo basic_info;
+  JxlEncoderInitBasicInfo(&basic_info);
+  basic_info.have_container = JXL_FALSE;
+  basic_info.xsize = width;
+  basic_info.ysize = height;
+  basic_info.bits_per_sample = 32;
+  basic_info.exponent_bits_per_sample = 8;
+  basic_info.intensity_target = 255.f;
+  basic_info.min_nits = 0.f;
+  basic_info.relative_to_max_display = JXL_FALSE;
+  basic_info.linear_below = 0.f;
+  basic_info.uses_original_profile = JXL_TRUE;
+  basic_info.have_preview = JXL_FALSE;
+  basic_info.have_animation = JXL_FALSE;
+  basic_info.orientation = JXL_ORIENT_IDENTITY;
+  basic_info.num_color_channels = 3;
+  basic_info.num_extra_channels = 0;
+  basic_info.alpha_bits = 0;
+  basic_info.alpha_exponent_bits = 0;
+  basic_info.alpha_premultiplied = JXL_FALSE;
+  JXL_ASSERT(JxlEncoderSetBasicInfo(encoder, &basic_info));
+
+  // We assume that the user wants the JXL image in a bmff container.
+  // JXL images can be stored without any container so they are smaller, but this
+  // removes the possibility of storing extra data like exif.
+  // It is not likely darktable users exporting images care so much about this
+  // small size improvement, and more likely they want to store exif data.
+  // HOWEVER - We do not want libjxl to write the image in a container, we will
+  // do that ourselves. This gives us the control to be able to add our own
+  // sections (e.g. exif data).
+  JXL_ASSERT(JxlEncoderUseContainer(encoder, JXL_FALSE));
+
+  // automatically freed when we destroy the encoder
+  JxlEncoderOptions *options = JxlEncoderOptionsCreate(encoder, NULL);
+  JXL_ASSERT(JxlEncoderOptionsSetLossless(options, params->lossless ? JXL_TRUE : JXL_FALSE));
+
+  int decode_effort = params->decoding_effort;
```

Comment:
Think it might be better to follow the API and cjxl command line options and actually use decode_speed as a direct parameter instead of decode_effort instead of this hidden acrobatics.